### PR TITLE
Metal: Add support for 2017 era iOS devices

### LINKED ISSUE #99682: iOS Mobile Renderer Black Screen
### Tested versions

- Reproduceable In: 4.4.dev.mono.custom_build.f0e971730

### System information

Godot v4.4.dev.mono (f0e971730) - macOS 15.0.1 - Multi-window, 1 monitor - Metal (Forward+) - integrated Apple M2 (Apple8) - Apple M2 (8 threads)

### Issue description

When exporting to iOS set to use the mobile renderer, it renders with a black screen:

Image

When exporting to iOS set to use the Forward+ renderer, it renders just fine:

Image

Here are the respective logs:

### Note

The Forward+ Renderer Log has many errors, but that is the one that actually renders correctly.  The Mobile Renderer log only has 1 error, but renders with a black screen.

### Mobile Renderer Log

```
Godot Engine v4.4.dev.mono.custom_build.f0e971730 (2024-11-16 22:16:24 UTC) - https://godotengine.org
Metal 3.1 - Forward Mobile - Using Device #0: Apple - Apple A12 GPU (Apple5)

ERROR: Mouse is not supported by this display server.
at: mouse_get_position (servers/display_server.cpp:508)
```

### Forward+ Renderer Log

[forward_plus_log.txt](https://github.com/user-attachments/files/17905110/forward_plus_log.txt)

### Steps to reproduce

1. Use Godot with double precision and mono support.
    - I'm not certain if this is necessary as I haven't tested other builds, but that is what I am currently using.
2. Open the MRP
3. Export to iOS
4. Install to an iPhone
5. See that a black screen renders
6. Change project settings mobile rendering method to Forward+
7. Export to IOS
8. Install to an iPhone
9. See that cube renders

### Minimal reproduction project (MRP)

[mrp.zip](https://github.com/user-attachments/files/17905053/mrp.zip)

Edit:

Here's the table of tested snapshots:

✅ = ~vulkan~ mobile works as expected
❌ = ~vulkan~ mobile blank screen

| Version  | Status  |
| -------- | ------- |
| 4.3.stable  | ✅ |
| 4.4-dev1 | ❌    |
| 4.4-dev2 | ?     |
| 4.4-dev3 | ?     |
| 4.4-dev4 | ?     |
| 4.4-dev5 | ❌    |

Edit:

I crossed out ~vulkan~ above.  @bruvzg pointed out [here](https://github.com/godotengine/godot/issues/99682#issuecomment-2500088341) that we are actually defaulting to the metal rendering device on ios now.  Changing it back to use Vulkan in the project settings as described in that comment makes the MRP work.  So it appears to be specifically related to the metal driver introduced in this pr: https://github.com/godotengine/godot/pull/88199

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #99682

Add support for Apple4 GPUs. 

Tier 2 argument buffers are required for read / write texture support, specifically for iOS, so this PR introduces support for generating shaders and binding without argument buffers for devices that only support Tier 1 argument buffers.

> [!IMPORTANT]
>
> I do not have an Apple4 GPU to test on, but I added an environment variable to allow me to force the shaders to be compiles without argument buffer support and was able to render the editor and complex scenes like the Bistro Demo.

### SPIRV-Cross Update

SPIRV-Cross is also updated to the latest, which includes fixes and support for mesh shaders, which will be used for future features.

### CODE REVIEW

**Comment by bruvzg** [comment]
> Tier 2 argument buffer support raises the minimum Apple GPU family from A11 (2017) to A13 (2019).

Should be fine and expected. We list A12 as min. supported for Vulkan. And I have already opened https://github.com/godotengine/godot-docs/pull/10325 to mention Metal and D3D12 with A13 listed as min. for Metal (based on Metal 3 supported devices page).

Old devices are supposed to use Compatibility renderer.

**Comment by TCROC** [comment]
> Cut off all iOS devices before 2019. Not sure what the market usage numbers are for older devices but 5 years for iOS devices isn't too crazy, people often praise them for their longevity.

I currently use the iPhone XR from 2018.  Which is what I discovered this issue on. @stuartcarnie is looking into a separate fix for older ones such as mine.

 > Have to make their game fallback properly to Compatibility, which kinda works for simple projects but may not work for the projects that would choose to use Forward+ or Mobile renderers.
 
Couldn't we just fall back to using Vulkan instead of Metal?  Instead of falling back to compatibility?  Vulkan works just fine on my iPhone XR.  Performance is even good with Forward+.

**Comment by stuartcarnie** [comment]
@akien-mga I'll have to implement a separate PR to disable argument buffers, as that will be a reasonable amount of work. Doing so will allow us to support mobile devices back to 2017 (as that is the earliest that supported image cube textures, required by Godot). 

Unfortunately, SPIRV-Cross does not work well (has bugs) with tier 1 argument buffers on iOS, when it has to fall back to a mix of classic and argument buffers if the pipeline contains writable textures.

**Comment by TCROC** [comment]
@stuartcarnie are we removing argument buffers entirely?  Or adding a boolean / flag to indicate whether the device supports it and fall back to something else when the device is old?

**Comment by stuartcarnie** [comment]
> @stuartcarnie are we removing argument buffers entirely? Or adding a boolean / flag to indicate whether the device supports it and fall back to something else when the device is old?

No, it will be a fallback for older devices, as argument buffers are more performant.

**Comment by TCROC** [comment]
@stuartcarnie Does this PR also fall back to vulkan for iOS devices older than 2017?

**Comment by TCROC** [comment]
Or is there a separate work in progress for that?

**Comment by stuartcarnie** [comment]
Vulkan won't work on devices prior to 2017 either, as Godot requires image cube map textures, which are only supported on 2017 devices and beyond.

**Comment by TCROC** [comment]
Gotcha.  Does it fall back to compatibility from metal then?

**Comment by stuartcarnie** [comment]
To put it another way, there should be no reason to require Vulkan / MoltenVK over Metal for Apple mobile devices, as they support the same range of devices.

For desktop, Vulkan / MoltenVK is required for x86 hardware.

**Comment by stuartcarnie** [comment]
> Gotcha. Does it fall back to compatibility from metal then?

Not sure – would have to ask the rendering team if that is the case.

**Review by akien-mga — APPROVED** [review]
Thirdparty changes seem fine to me, and yay for more hardware support!

File: drivers/metal/rendering_device_driver_metal.h [hunk 1]
```diff
@@ -61,7 +61,7 @@ class API_AVAILABLE(macos(11.0), ios(14.0)) RenderingDeviceDriverMetal : public
 
 	uint32_t version_major = 2;
 	uint32_t version_minor = 0;
-	MetalDeviceProperties *metal_device_properties = nullptr;
+	MetalDeviceProperties *device_properties = nullptr;
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.h:64]: I dropped the `metal_` prefix, as it is obviously for `metal` 🤦🏻

File: drivers/metal/rendering_device_driver_metal.mm [hunk 2]
```diff
@@ -1058,7 +1058,7 @@ static const API_AVAILABLE(macos(11.0), ios(14.0)) MTLSamplerBorderColor SAMPLER
 
 #pragma mark - Shader
 
-const uint32_t SHADER_BINARY_VERSION = 3;
+const uint32_t SHADER_BINARY_VERSION = 4;
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:1061]: We have to include a flag to indicate whether the shader was compiled to generate an ABI with argument buffers or classic binding model.

File: drivers/metal/rendering_device_driver_metal.mm [hunk 3]
```diff
@@ -1536,15 +1539,19 @@ void deserialize(BufReader &p_reader) {
...
+	};
 	CharString shader_name;
 	// The Metal language version specified when compiling SPIR-V to MSL.
 	// Format is major * 10000 + minor * 100 + patch.
 	uint32_t msl_version = UINT32_MAX;
 	uint32_t vertex_input_mask = UINT32_MAX;
 	uint32_t fragment_output_mask = UINT32_MAX;
 	uint32_t spirv_specialization_constants_ids_mask = UINT32_MAX;
-	uint32_t is_compute = UINT32_MAX;
-	uint32_t needs_view_mask_buffer = UINT32_MAX;
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:1547]: I dropped `is_compute` as we don't need to store it, and moved any boolean flags to a single `uint32_t`, so it is easier to add them in the future.

File: drivers/metal/rendering_device_driver_metal.mm [hunk 4]
```diff
@@ -1999,10 +2032,25 @@ void deserialize(BufReader &p_reader) {
...
 #if TARGET_OS_IOS
-	msl_options.ios_use_simdgroup_functions = (*metal_device_properties).features.simdPermute;
+	msl_options.ios_use_simdgroup_functions = (*device_properties).features.simdPermute;
 #endif
 
-	msl_options.argument_buffers = true;
+	bool disable_argument_buffers = false;
+	if (String v = OS::get_singleton()->get_environment(U"GODOT_DISABLE_ARGUMENT_BUFFERS"); v == U"1") {
+		disable_argument_buffers = true;
+	}
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:2041]: It's undocumented, and I use this for debugging purposes. I was able to test out argument buffers for a number of projects and they ran without issue

File: drivers/metal/rendering_device_driver_metal.mm [hunk 5]
```diff
@@ -2042,7 +2090,12 @@ void deserialize(BufReader &p_reader) {
 		std::unordered_set<VariableID> active = compiler.get_active_interface_variables();
 		ShaderResources resources = compiler.get_shader_resources();
 
-		std::string source = compiler.compile();
+		std::string source;
+		try {
+			source = compiler.compile();
+		} catch (CompilerError &e) {
+			ERR_FAIL_V_MSG(Result(), "Failed to compile stage " + String(SHADER_STAGE_NAMES[stage]) + ": " + e.what());
+		}
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:2098]: It's possible that `compile` will throw an exception

File: drivers/metal/rendering_device_driver_metal.mm [hunk 6]
```diff
@@ -2088,8 +2141,8 @@ void deserialize(BufReader &p_reader) {
 			return res;
 		};
 
-		auto descriptor_bindings = [&compiler, &active, &uniform_sets, stage, &get_decoration](SmallVector<Resource> &resources, Writable writable) {
-			for (Resource const &res : resources) {
+		auto descriptor_bindings = [&compiler, &active, &uniform_sets, stage, &get_decoration](SmallVector<Resource> &p_resources, Writable p_writable) {
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:2144]: I renamed the arguments of the lambda to include the `p_` prefix

File: drivers/metal/rendering_device_driver_metal.mm [hunk 7]
```diff
@@ -2384,6 +2437,11 @@ void deserialize(BufReader &p_reader) {
 			ERR_FAIL_V_MSG(ShaderID(), "Unexpected end of buffer");
 	}
 
+	// We need to regenerate the shader if the cache is moved to an incompatible device.
+	ERR_FAIL_COND_V_MSG(device_properties->features.argument_buffers_tier < MTLArgumentBuffersTier2 && binary_data.uses_argument_buffers(),
+			ShaderID(),
+			"Shader was generated with argument buffers, but device has limited support");
+
```
Comment (stuartcarnie) [inline drivers/metal/rendering_device_driver_metal.mm:2444]: This ensures that if someone copies the Godot `shader_cache` to a device that supports a lower argument buffer tier, that it won't crash, but regenerate the shaders.

File: drivers/metal/metal_objects.mm [hunk 8]
```diff
@@ -249,7 +249,7 @@
 	const MDSubpass &subpass = render.get_subpass();
 
 	uint32_t vertex_count = p_rects.size() * 6 * subpass.view_count;
-	simd::float4 vertices[vertex_count];
+	simd::float4 *vertices = ALLOCA_ARRAY(simd::float4, vertex_count);
```
Comment (stuartcarnie) [inline drivers/metal/metal_objects.mm:252]: Annoying compiler warning in my IDE, so I switched to `alloca`

File: drivers/metal/metal_objects.mm [hunk 9]
```diff
@@ -1093,9 +1047,373 @@
 	}
 }
 
-BoundUniformSet &MDUniformSet::boundUniformSetForShader(MDShader *p_shader, id<MTLDevice> p_device) {
+void MDUniformSet::bind_uniforms_argument_buffers(MDShader *p_shader, MDCommandBuffer::RenderState &p_state) {
```
Comment (stuartcarnie) [inline drivers/metal/metal_objects.mm:1050]: This is the original argument buffers implementation moved to a new method, so we can call the right binding, depending on whether the shader uses argument buffer ABI or classic ABI

**Review by clayjohn — APPROVED** [review]
Looks fine to me. I don't know much about the Metal API  and I don't have a 2017-era device to test with, so TIWAGOS

**Comment by Repiteo** [comment]
Thanks!

**Comment by TCROC** [comment]
@stuartcarnie I got around to testing this in my fork.  It appears to fix the black screen!  However, I encountered this new error:

```
validateComputeFunctionArguments:1083: failed assertion `Compute Function(main0): argument src_particles[0] from buffer(4) with offset(0) and length(16) has space for 16 bytes, but argument has a length(128).'
```

It doesn't seem to have any impact that I can yet notice.  It may be something I did wrong in my fork as well.  I'll test with vanilla godot master to see if anything changes.  Thank you again for all of the hard work you did on this!

**Comment by stuartcarnie** [comment]
@TCROC thanks for reporting your issue; I've created #100764 to track it, as I have reproduced it. Fix coming soon.

**Comment by TCROC** [comment]
Thanks @stuartcarnie ! I look forward to testing it out! :)

**Pull request merged**